node/src/components/contract_runtime.rs

```diff
@@ -202,362 +298,341 @@ where
+                            let result = engine_state.read_trie(correlation_id, trie_key);
+                            metrics.read_trie.observe(start.elapsed().as_secs_f64());
+                            let result = match result {
+                                Ok(result) => result,
+                                Err(error) => {
+                                    error!(?error, "read_trie_request");
+                                    None
                                 }
-                                Err(GetEraValidatorsError::EraValidatorsMissing) => Ok(None),
-                                Err(error) => Err(error),
                             };
-                        metrics
-                            .get_era_validator_weights_by_era_id
-                            .observe(start.elapsed().as_secs_f64());
-                        ret
-                    })
-                    .await
-                    .expect("should run");
-                    trace!(?result, "get validator weights by era id response");
-                    responder.respond(result).await
+                            trace!(?result, "read_trie response");
+                            responder.respond(result).await
+                        }
+                        .ignore()
+                    }
+                    ContractRuntimeRequest::PutTrie { trie, responder } => {
+                        trace!(?trie, "put_trie request");
+                        let engine_state = Arc::clone(&self.engine_state);
+                        let metrics = Arc::clone(&self.metrics);
+                        async move {
+                            let correlation_id = CorrelationId::new();
+                            let start = Instant::now();
+                            let result = engine_state
+                                .put_trie_and_find_missing_descendant_trie_keys(
+                                    correlation_id,
+                                    &*trie,
+                                );
+                            metrics.put_trie.observe(start.elapsed().as_secs_f64());
+                            trace!(?result, "put_trie response");
+                            responder.respond(result).await
+                        }
+                        .ignore()
+                    }
+                    ContractRuntimeRequest::ExecuteBlock(finalized_block) => {
+                        debug!(?finalized_block, "execute block");
+                        effect_builder
+                            .get_block_at_height_local(finalized_block.height())
+                            .event(move |maybe_block| {
+                                maybe_block.map(Box::new).map_or_else(
+                                    || Event::BlockIsNew(Box::new(finalized_block)),
+                                    Event::BlockAlreadyExists,
+                                )
+                            })
```

Comment:
If you put it inside the `async move` block then we should get away without the event?